### PR TITLE
enable users recover without password

### PR DESCRIPTION
**Pull request by abramsymons**
This pull request enable users to recover their account if they do not have a backup of their connections/groups names and photos or they lost their backup password.

### New process
1) Recovering user create an AES key and share it via QR code to the recovery connection to be used to encrypt and decrypt sensitive data that will be transferred 
2) sha256 hash of AES key will be used as channelId to transfer data between Recovering user and recovery connection
3) Recovering user initiate the process upload new singing key and timestamp to the channel using `data` as `dataId`
4) Recovery connection scan the AES key from the QR code, calculate `channelId`, and download `data` as the signature request, sign the request and upload it to the channel using `sig_{his_own_brightid}`
5) Recovery connection get list users who connected to the recovering user with `just_met`, `already_known` and `recovery` and check his own connections list and then encrypt and upload names and photos of mutual ones to the channel if those did not have already uploaded by another recovery connection.
6) Recovery connection get list of groups of recovering user and check his own groups and then encrypt and upload names and photos of mutual groups to the channel if those did not have already uploaded by another recovery connection.
7) Recovering user constantly check the channel and download connections/groups info beside signatures and add those info to its store
8) Recovering user stop checking the channel when 2 signatures exist and no new data be uploaded for 5 seconds and move user to restore screen after that.
9) Recovering users can select "skip" when they will be asked to enter their password and if they did so, the client restore their account with connections/groups info that their recovery connections provided.

### CODE REVIEW

**Comment by adamstallard** [comment]
I understand it except for these steps:

>3. Recovering user initiate the process upload new singing key and timestamp to the channel using data as channelId
>4. Recovery connection scan the AES key from the QR code, calculate channelId, and download data as the signature request, sign the request and upload it to the channel using sig_{his_own_brightid}

What does "using data as channelId" mean?  I think 3. has to do with uploading the new signing key public key and a timestamp for the recovery connection to sign and return in step 4. , but I don't know what "using data as channelId" means. Wouldn't the channelId be the hash of the AES key that was in the QR code?

I think 4. has to do with the recovery connection uploading one of the signatures needed for the `setSigningKey` operation. Is that right?

**Comment by abramsymons** [comment]
> What does "using data as channelId" mean?

Sorry, I mean using `data` as `dataId`. We send sigs, connections and groups info to same channel using different `dataId`s and we use `data` as `dataId` for the item recovering user upload to channel as his request.

> I think 4. has to do with the recovery connection uploading one of the signatures needed for the setSigningKey operation. Is that right?

Yes, and this time, recovery connection use `sig_{his_own_brightid}` as `dataId` when sending signature to the channel.

**Review by adamstallard — APPROVED** [review]
This is a big improvement. Nice work.

File: BrightID/src/components/Recovery/RecoveringConnectionCard.js [hunk 1]
```diff
@@ -25,24 +30,24 @@ class RecoveryConnectionCard extends React.PureComponent<Props> {
...
-        t('common.alert.info'),
-        t('restore.alert.text.requestRecovering'),
-        [
-          {
-            text: t('common.alert.ok'),
-            onPress: () => this.props.navigation.navigate('Home'),
-          },
-        ],
-      );
+      const ipAddress = await api.ip();
```
Comment (TripleSpeeder) [inline BrightID/src/components/Recovery/RecoveringConnectionCard.js:35]: There is a lot of network activity going on in this handler without visible user feedback. On bad network condition or when having much data to transfer the user will not know if the app has crashed or is still working.

I think we need:
- Button should be disabled after pressed, so user can not trigger multiple parallel executions
- Screen should show a waitcursor/spinner, so it is clear for the user that something is going on (Best case would be a nice progress bar, possibly reusing `Tasks/TasksProgress.js`?)

**Review by TripleSpeeder — CHANGES_REQUESTED** [review]
This will be an awesome feature! But I think the UI side needs some improvement, see my comment on RecoveringConnectionCard.

Also i notice the code does not implement the eslint formatting rules. Can you run `eslint --fix` on the files you changed before commiting? I have this automated in my IDE - whenever I save a file it automatically gets linted :-)

File: BrightID/src/components/Recovery/RecoveringConnectionCard.js (hunk 1 above)
Comment (abramsymons) [reply BrightID/src/components/Recovery/RecoveringConnectionCard.js:35]: We already have a spinner added there that shows waiting to users while their app is uploading the mutual info.
Disabling the button is a good idea too.

**Comment by abramsymons** [comment]
> Also i notice the code does not implement the eslint formatting rules. Can you run `eslint --fix` on the files you changed before commiting? I have this automated in my IDE - whenever I save a file it automatically gets linted :-)

Sorry, I did not know `eslint`, thank you for informing me about that :)
I will use that for future commits.

**Pull request merged**